src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,511 @@
+            // E.g. when the user triggers completion for an if statement ($$ is the caret location) we insert braces to get
+            // if (true){$$}
+            // We then need to format this to
+            // if (true) { $$}
+            var (formattingChanges, finalCurlyBraceEnd) = await FormatTrackingSpanAsync(
+                braceCompletionContext.Document,
+                documentOptions,
+                braceCompletionContext.OpeningPoint,
+                braceCompletionContext.ClosingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: true,
+                // We're not trying to format the indented block here, so no need to pass in additional rules.
+                braceFormattingIndentationRules: ImmutableArray<AbstractFormattingRule>.Empty,
+                cancellationToken).ConfigureAwait(false);
+
+            if (formattingChanges.IsEmpty)
+            {
+                return null;
+            }
+
+            // The caret location should be at the start of the closing brace character.
+            var originalText = await braceCompletionContext.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var formattedText = originalText.WithChanges(formattingChanges);
+            var caretLocation = formattedText.Lines.GetLinePosition(finalCurlyBraceEnd - 1);
+
+            return new BraceCompletionResult(formattingChanges, caretLocation);
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext context, CancellationToken cancellationToken)
+        {
+            var document = context.Document;
+            var closingPoint = context.ClosingPoint;
+            var openingPoint = context.OpeningPoint;
+            var originalDocumentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // check whether shape of the braces are what we support
+            // shape must be either "{|}" or "{ }". | is where caret is. otherwise, we don't do any special behavior
+            if (!ContainsOnlyWhitespace(originalDocumentText, openingPoint, closingPoint))
+            {
+                return null;
+            }
+
+            // If there is not already an empty line inserted between the braces, insert one.
+            TextChange? newLineEdit = null;
+            var textToFormat = originalDocumentText;
+            if (!HasExistingEmptyLineBetweenBraces(openingPoint, closingPoint, originalDocumentText))
+            {
+                newLineEdit = new TextChange(new TextSpan(closingPoint - 1, 0), Environment.NewLine);
+                textToFormat = originalDocumentText.WithChanges(newLineEdit.Value);
+            }
+
+            // Modify the closing point location to adjust for the newly inserted line.
+            closingPoint += Environment.NewLine.Length;
```

Comment:
don't love this.  it seems like it would actually be wrong in many circumstances (for example .editorconfig setting newline to '\n').  deserves tests of that :)